lib/rucio/common/stomp_utils.py

```diff
@@ -15,145 +15,415 @@
+        """
+        return self._brokers
+
+
+class ListenerBase(HeartbeatListener):
+    """Listener Base."""
 
-    LoggerFunction = Callable[..., Any]
+    _logger = formatted_logger(logging.log, 'ListenerBase %s')
 
+    def __init__(self,
+                 conn: Connection,
+                 logger: None | LoggerFunction = None,
+                 **kwargs):
+        """
+        Initialise.
 
-def resolve_ips(fqdns: "Sequence[str]", logger: "LoggerFunction" = logging.log):
-    logger(logging.DEBUG, 'resolving dns aliases: %s' % fqdns)
-    resolved = []
-    for fqdn in fqdns:
-        addrinfos = socket.getaddrinfo(fqdn, 0, socket.AF_INET, 0, socket.IPPROTO_TCP)
-        resolved.extend(ai[4][0] for ai in addrinfos)
-    logger(logging.DEBUG, 'dns aliases resolved to %s', resolved)
-    return resolved
+        Args:
+            conn (Connection12): The connection object that is using this listener
+            logger (logging.Logger): Logger to use. Defaults to logging.getLogger(__name__).getChild(__qualname__).
+
+        Kwargs:
+            Arguments to pass to the stomp.ConnectionListener base class.
+        """
+        super().__init__(transport=conn.transport, **kwargs)
+        self._conn = conn
+        if logger is not None:
+            self._logger = logger
+
+    @METRICS.count_it
+    def on_heartbeat_timeout(self):
+        self._conn.disconnect()
+
+    @METRICS.count_it
+    def on_error(self, frame: Frame):
+        """
+        on_error
+        """
+        self._logger(logging.ERROR, 'Message receive error: [%s] %s', self._conn.brokers[0][0], frame.body)
+
+
+StompConfig = namedtuple("StompConfig", ('brokers', 'use_ssl', 'port', 'vhost',
+                                         'destination', 'key_file', 'cert_file',
+                                         'username', 'password', 'nonssl_port',
+                                         'reconnect_attempts_max', 'timeout', 'heartbeats'))
 
 
 class StompConnectionManager:
+    """Stomp Connection Manager."""
 
-    def __init__(self):
-        self._brokers = None
-        self._port = None
-        self._use_ssl = None
-        self._vhost = None
-        self._reconnect_attempts = None
-        self._ssl_key_file = None
-        self._timeout = None
-        self._heartbeats = None
+    _logger = formatted_logger(logging.log, 'StompConnectionManager %s')
 
-        self._connections = {}
+    def __init__(self,
+                 config_section: str,
+                 logger: None | LoggerFunction = None):
+        """
+        Initialise.
 
-    def is_stalled(self, connection: Connection, *, logger: "LoggerFunction" = logging.log):
-        if not connection.is_connected():
-            return True
+        Args:
+            config_section (str): The name of the config section for this manager to parse for configuration.
+            logger (logging.Logger): logger to use. Defaults to logging.getLogger(__name__).getChild(__qualname__).
```

Comment:
Please avoid specifying the argument types in the docstring if they are already present as type annotations; there's no need to duplicate this information and it can cause mismatches between the type annotation and the type specified in the docstring if changes are made down the line